carla/recourse_methods/catalog/cem/cem.py

```diff
@@ -0,0 +1,578 @@
+            )
+            Loss_Attack_s = tf.maximum(
+                0.0, max_nontarget_lab_score_s - target_lab_score_s + self.kappa
+            )
+        elif self.mode == "PN":
+            Loss_Attack = tf.maximum(
+                0.0, -self.max_nontarget_lab_score + self.target_lab_score + self.kappa
+            )
+            Loss_Attack_s = tf.maximum(
+                0.0, -max_nontarget_lab_score_s + target_lab_score_s + self.kappa
+            )
+
+        # sum up the losses
+        self.Loss_L1Dist = tf.reduce_sum(self.L1_dist)
+        self.Loss_L1Dist_s = tf.reduce_sum(self.L1_dist_s)
+
+        self.Loss_L2Dist = tf.reduce_sum(self.L2_dist)
+        self.Loss_L2Dist_s = tf.reduce_sum(self.L2_dist_s)
+
+        self.Loss_Attack = tf.reduce_sum(self.const * Loss_Attack)
+        self.Loss_Attack_s = tf.reduce_sum(self.const * Loss_Attack_s)
+
+        if self.mode == "PP":
+            self.Loss_AE_Dist = self.gamma * tf.square(
+                tf.norm(self.AE(self.delta_img) - self.delta_img)
+            )
+            self.Loss_AE_Dist_s = self.gamma * tf.square(
+                tf.norm(self.AE(self.delta_img) - self.delta_img_s)
+            )
+        elif self.mode == "PN":
+            self.Loss_AE_Dist = self.gamma * tf.square(
+                tf.norm(self.AE(self.adv_img) - self.adv_img)
+            )
+            self.Loss_AE_Dist_s = self.gamma * tf.square(
+                tf.norm(self.AE(self.adv_img_s) - self.adv_img_s)
+            )
+
+        self.Loss_ToOptimize = (
+            self.Loss_Attack_s + self.Loss_L2Dist_s + self.Loss_AE_Dist_s
+        )
+        self.Loss_Overall = (
+            self.Loss_Attack
+            + self.Loss_L2Dist
+            + self.Loss_AE_Dist
+            + tf.multiply(self.beta, self.Loss_L1Dist)
+        )
+
+        self.learning_rate = tf.train.polynomial_decay(
+            self.INIT_LEARNING_RATE, self.global_step, self.MAX_ITERATIONS, 0, power=0.5
+        )
+        optimizer = tf.train.GradientDescentOptimizer(self.learning_rate)
+        start_vars = set(x.name for x in tf.global_variables())
+        self.train = optimizer.minimize(
+            self.Loss_ToOptimize,
+            var_list=[self.adv_img_s],
+            global_step=self.global_step,
+        )
+        end_vars = tf.global_variables()
+        new_vars = [x for x in end_vars if x.name not in start_vars]
+
+        # these are the variables to initialize when we run
+        self.setup = []
+        self.setup.append(self.orig_img.assign(self.assign_orig_img))
+        self.setup.append(self.target_lab.assign(self.assign_target_lab))
+        self.setup.append(self.const.assign(self.assign_const))
+        self.setup.append(self.adv_img.assign(self.assign_adv_img))
+        self.setup.append(self.adv_img_s.assign(self.assign_adv_img_s))
+
+        self.init = tf.variables_initializer(
+            var_list=[self.global_step] + [self.adv_img_s] + [self.adv_img] + new_vars
+        )
```

Comment:
Not really, this is how the method initializes itself. Like I could put everything in `other_init()` or something, but that doesn't seem much better.

Reply:
Refactoring could also mean extracting some parts into new methods, which mostly eliminates duplicate code and makes it more readable. 
https://makolyte.com/refactoring-the-long-method-code-smell/

I had to do the same for FACE, which made it by far easier to understand.

Reply:
I'm don't think there is much to gain here in that respect. There isn't really duplicate code. If you have some pointers I would be happy to change it though.

Reply:
So I did a bit of refactoring and was able to collapse `__init__()` for more than 100 lines of code just by extracting methods. There was plenty of duplicate code, for example all conditions (`cond1`, ..., `cond7`) could be summarized into one single method.

But this constructor is still to big and not readable, plus there should be more refactor potential. For example, variable initialization for `assign_adv_img `and `assign_adv_img_s`, or `delta_img `and `delta_img_s`, `...`
Maybe @indyfree knows some best practices for such cases.

It would be great if you could have another look into that, and try the same with the `attack()` method. I'll try to do the same in the next couple of days.

Reply:
made issue.